src/ratelimit/RateLimitController.ts

```diff
@@ -0,0 +1,188 @@
+import type { Patch } from 'immer';
+
+import { BaseController } from '../BaseControllerV2';
+
+import type { RestrictedControllerMessenger } from '../ControllerMessenger';
+
+/**
+ * @type RateLimitState
+ * @property requests - Object containing number of requests in a given interval for each origin and api type combination
+ */
+export type RateLimitState = {
+  requests: Record<ApiType, Record<string, number>>;
+};
+
+export enum ApiType {
+  showNativeNotification = 'showNativeNotification',
+}
+
+export interface CallArgs {
+  /**
+   * Enum type to determine API type.
+   */
+  type: ApiType;
+
+  /**
+   * Args passed to the API call
+   */
+  // @todo Type this?
+  args: any;
+}
+
+const name = 'RateLimitController';
+
+export type RateLimitStateChange = {
+  type: `${typeof name}:stateChange`;
+  payload: [RateLimitState, Patch[]];
+};
+
+export type GetRateLimitState = {
+  type: `${typeof name}:getState`;
+  handler: () => RateLimitState;
+};
+
+export type CallAPI = {
+  type: `${typeof name}:call`;
+  handler: RateLimitController['call'];
+};
+
+export type ControllerActions = GetRateLimitState | CallAPI;
+
+export type RateLimitMessenger = RestrictedControllerMessenger<
+  typeof name,
+  ControllerActions,
+  RateLimitStateChange,
+  never,
+  never
+>;
+
+const metadata = {
+  requests: { persist: false, anonymous: false },
+};
+
+const defaultState = {
+  requests: { [ApiType.showNativeNotification]: {} },
+};
+
+/**
+ * Controller that handles showing notifications to the user and rate limiting origins
+ */
+export class RateLimitController extends BaseController<
+  typeof name,
+  RateLimitState,
+  RateLimitMessenger
+> {
+  private showNativeNotification;
+
+  private rateLimitTimeout;
+
+  private rateLimitCount;
+
+  /**
+   * Creates a RateLimitController instance.
+   *
+   * @param options - Constructor options.
+   * @param options.messenger - A reference to the messaging system.
+   * @param options.state - Initial state to set on this controller.
+   * @param options.showNativeNotification - Function that shows a native notification in the consumer
+   * @param options.rateLimitTimeout - The time window in which the rate limit is applied
+   * @param options.rateLimitCount - The amount of notifications an origin can show in the rate limit time window
+   */
+  constructor({
+    // @todo Pick some sane defaults for this
+    rateLimitTimeout = 5000,
+    rateLimitCount = 3,
+    messenger,
+    state,
+    showNativeNotification,
+  }: {
+    rateLimitTimeout?: number;
+    rateLimitCount?: number;
+    messenger: RateLimitMessenger;
+    state?: Partial<RateLimitState>;
+    showNativeNotification: (
+      title: string,
+      message: string,
+      url?: string,
+    ) => void;
+  }) {
+    super({
+      name,
+      metadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this.showNativeNotification = showNativeNotification;
```

Comment:
We could make this controller generic over the actions being rate limited, and pass them in as a mapping of type to implementation.

That would let us extend what is rate limited without altering the controller itself, and would decouple this controller from the things being rate limited.

Reply:
Done!